### ISSUE TITLE
Misleading error message when accidently trying to upload a dictionary to blob storage

### ISSUE THREAD

**Issue by edwinvehmaanpera** [Azure/azure-sdk-for-python#18122]
- **Package Name**: azure-storage-blob
- **Package Version**: 12.8.0
- **Operating System**: Windows 10
- **Python Version**:  Python 3.8.8

**Describe the bug**
Accidentally passed a dictionary as data to upload to blob storage. The error message:

 ```
azure.core.exceptions.HttpResponseError: Operation returned an invalid status 'Server failed to authenticate the request. 
Make sure the value of Authorization header is formed correctly including the signature.'
 ```
I hadn't noticed the data was a dictionary and the error message confused me. I troubleshooted for a good while to find what  my "authentication" error might be. 

**To Reproduce**
1. Run code like:
 ```python
from azure.storage.blob import BlobServiceClient, BlobClient, ContainerClient

connect_str = '<Connection String>'

blob_service_client = BlobServiceClient.from_connection_string(connect_str)

blob_client = blob_service_client.get_blob_client(container='test', blob='test.txt')
data = {"data":"test"}
blob_client.upload_blob(data)
```

**Expected behavior**
Check what gets passed as data to upload_blob and give a more descriptive and accurate error message or upload the dictionary as a string. There might be similar issues for other data types/sctructures as well.

**Comment by mccoyp** [comment]
Hi @edwinvehmaanpera, thanks for opening this issue! We'll investigate ASAP.

**Comment by tasherif-msft** [comment]
Thank you for pointing this out. I will look into improving our type annotations and/or our thrown exception.

**Comment by vincenttran-msft** [comment]
Hi @edwinvehmaanpera Edwin, thanks for your inquiry. I have been able to successfully reproduce this behavior and will work with the team to improve the thrown exception. I will post an update as soon as a resolution is met!
Thanks!

**Comment by vincenttran-msft** [comment]
This issue has been addressed and resolved in #24730. Thanks again for bringing awareness to this issue!
